pkg/webhook/common.go

```diff
@@ -148,11 +149,23 @@ func (h *webhookHandler) skipExcludedNamespace(req *admissionv1.AdmissionRequest
 	return isNamespaceExcluded, err
 }
 
-func getServerConfig(mgr manager.Manager) *webhook.Server {
-	server := mgr.GetWebhookServer()
+func congifureWebhookServer(server *webhook.Server) *webhook.Server {
 	server.TLSMinVersion = *tlsMinVersion
 	if *clientCAName != "" {
 		server.ClientCAName = *clientCAName
+		server.TLSOpts = []func(*tls.Config){
+			func(cfg *tls.Config) {
+				cfg.VerifyConnection = func(cs tls.ConnectionState) error {
+					if len(cs.PeerCertificates) > 0 {
+						if cs.PeerCertificates[0].Subject.CommonName != *certCNName {
```

Comment:
Is it correct to only check the first PeerCertificate in the list?

Reply:
There is only one entry in `PeerCertificate`. When API server talks to gatekeeper webhook, it will pass only one cert as configured via `--admission-config` file. So checking the first entry would work as there won't be any other cert in the array.

Reply:
> There is only one entry in `PeerCertificate`. When API server talks to gatekeeper webhook, it will pass only one cert as configured via `--admission-config` file. So checking the first entry would work as there won't be any other cert in the array.

We do not need to make assumptions about the client.  It can be anything, and it can send any number of intermediate certs that chain up to the root.

```go
	// PeerCertificates are the parsed certificates sent by the peer, in the
	// order in which they were sent. The first element is the leaf certificate
	// that the connection is verified against.
	//
	// On the client side, it can't be empty. On the server side, it can be
	// empty if Config.ClientAuth is not RequireAnyClientCert or
	// RequireAndVerifyClientCert.
	PeerCertificates []*x509.Certificate

	// VerifiedChains is a list of one or more chains where the first element is
	// PeerCertificates[0] and the last element is from Config.RootCAs (on the
	// client side) or Config.ClientCAs (on the server side).
	//
	// On the client side, it's set if Config.InsecureSkipVerify is false. On
	// the server side, it's set if Config.ClientAuth is VerifyClientCertIfGiven
	// (and the peer provided a certificate) or RequireAndVerifyClientCert.
	VerifiedChains [][]*x509.Certificate
```

The important part is that on the server side:

```
	// PeerCertificates ...
	// ... The first element is the leaf certificate
	// that the connection is verified against.
```